README.md
SparseSignals
=============


NEWS
TODO

Vignette explaining when we get speedups vs just computing with a
repetitive piecewise constant vector.

2014.3.12

getVector for making a piecewise constant vector from a SparseSignal.

return SparseSignal invisibly after printing.

in add_SparseSignals_interface, allocate let
z_size_max=2*(x_size+y_size) instead of only x_size+y_size.

2014.3.11

S3 methods for printing, addition, and subtraction.

First version, interface to add_SparseSignals.

tests/testthat.R
library(testthat)
library(SparseSignals)
test_package("SparseSignals")


inst/tests/test-add.R
context("add two SparseSignals")

test_that("adding nonsense gives errors", {
  x <- SparseSignal(c(12L, 10L, 14L),
                    c(13L, 11L, 16L),
                    c( 2,  1,  3))
  y <- SparseSignal(c(10L, 12L, 15L, 1000L),
                    c(11L, 13L, 17L, 1001L),
                    c( 1, -2,  4,    0))
  expect_that(addSparseSignals(x, y), throws_error("positions decreasing"))
})

test_that("adding gives the right answer", {
  x <- data.frame(1L, 2L, 5.5)
  two <- addSparseSignals(x, data.frame(4L, 10L, 3.8))
  expect_that(two$first, equals(c(1, 4)))
  expect_that(two$after, equals(c(2, 10)))
  expect_that(two$value, equals(c(5.5, 3.8)))

  neg <- data.frame(1L, 2L, -5.5)
  none <- addSparseSignals(neg, x)
  expect_equal(length(none$first), 0)
  expect_equal(length(none$after), 0)
  expect_equal(length(none$value), 0)

  three <-
    addSparseSignals(data.frame(1L, 5L, 5.5),
                     data.frame(3L, 10L, 3.3))
  expect_that(three$first, equals(c(1, 3, 5)))
  expect_that(three$after, equals(c(3, 5, 10)))
  expect_that(three$value, equals(c(5.5, 8.8, 3.3)))

  x <- data.frame(c(10L, 12L, 14L),
                  c(11L, 13L, 16L),
                  c(  1,   2,   3))
  y <- data.frame(c(10L, 12L, 15L),
                  c(11L, 13L, 17L),
                  c( 1, -2,    4))
  z <- addSparseSignals(x, y)
  expect_equal(z$first, c(10, 14, 15, 16))
  expect_equal(z$after, c(11, 15, 16, 17))
  expect_equal(z$value, c( 2,  3,  7,  4))
})


inst/tests/test-methods.R
context("methods")

test_that("adding nonsense gives errors", {
  x <- SparseSignal(c(10L, 12L, 14L),
                    c(11L, 13L, 16L),
                    c(  1,   2,   3))
  expect_error(x + c(1,2))
  expect_error(x - c(1,2))
  for(nonsense in list(list(), function(x)x)){
    expect_error(x + nonsense, 
                 "addition only defined when y is numeric or SparseSignal")
    expect_error(x - nonsense,
                 "subtraction only defined when y is numeric or SparseSignal")
  }
})

test_that("SparseSignal + vector gives the right answer", {
  x <- SparseSignal(c(10L, 12L, 14L),
                    c(11L, 13L, 16L),
                    c(  1,   2,   3))
  plus.scalar <- x + 1
  expect_that(plus.scalar$value, equals(2:4))
  plus.vector <- x + 3:1
  expect_that(plus.vector$value, equals(c(4, 4, 4)))
  minus <- -x$value
  expect_true(nrow(x + minus) == 0)
})

test_that("SparseSignal - vector gives the right answer", {
  x <- SparseSignal(c(10L, 12L, 14L),
                    c(11L, 13L, 16L),
                    c(  1,   2,   3))
  minus.scalar <- x - 10
  expect_that(minus.scalar$value, equals((-9):(-7)))
  minus.vector <- x - c(2, 3, 4)
  expect_that(minus.vector$value, equals(c(-1, -1, -1)))
  expect_true(nrow(x - x$value) == 0)
})


src/main.c
/* -*- compile-command: "gcc main.c SparseSignals.c -o test && ./test" -*- */
#include <stdio.h>
#include <stdlib.h>
#include "SparseSignals.h"
int main(void){
    int x_first[]={10,  150};
    int x_after[]={100, 200};
    int y_first[]={10, 190};
    int y_after[]={160, 250};
    double x_value[]={5.5, 5.5};
    double y_value[]={-5.5, 10};
    int x_size=2, y_size=2, i;
    int z_size_max = x_size + y_size;
    int    *z_first = (int*)    malloc(z_size_max * sizeof(int));
    int    *z_after = (int*)    malloc(z_size_max * sizeof(int));
    double *z_value = (double*) malloc(z_size_max * sizeof(double));
    int z_size = add_SparseSignals(
	x_first, x_after, x_value, x_size,
	y_first, y_after, y_value, y_size,
	z_first, z_after, z_value);
    for(i=0; i<z_size;i++){
	printf("%4d %4d %f\n", z_first[i], z_after[i], z_value[i]);
    }
    return 0;
}


src/interface.c
/* -*- compile-command: "R CMD INSTALL .. && R CMD check .." -*- */
#include <R.h>
#include <Rinternals.h>
#include "SparseSignals.h"

SEXP 
add_SparseSignals_interface
(SEXP x, SEXP y){
    SEXP z, class, names, first, after, value;
    int    *x_first = INTEGER(VECTOR_ELT(x, 0));
    int    *x_after = INTEGER(VECTOR_ELT(x, 1));
    double *x_value =    REAL(VECTOR_ELT(x, 2));
    int    *y_first = INTEGER(VECTOR_ELT(y, 0));
    int    *y_after = INTEGER(VECTOR_ELT(y, 1));
    double *y_value =    REAL(VECTOR_ELT(y, 2));

    int x_size = length(VECTOR_ELT(x, 0)), y_size = length(VECTOR_ELT(y, 0));
    int z_size_max = (x_size + y_size)*2;
    int    *z_first = (int*)    R_alloc(z_size_max, sizeof(int));
    int    *z_after = (int*)    R_alloc(z_size_max, sizeof(int));
    double *z_value = (double*) R_alloc(z_size_max, sizeof(double));
    //Rprintf("z_size_max=%d\n", z_size_max);
    int z_size = add_SparseSignals(
	x_first, x_after, x_value, x_size,
	y_first, y_after, y_value, y_size,
	z_first, z_after, z_value);
    //Rprintf("z_size=%d\n", z_size);
    if(z_size == ERROR_POSITIONS_DECREASING)error("positions decreasing");
    if(z_size < 0)error("undefined error code %d", z_size);

    PROTECT(z = allocVector(VECSXP, 3));

    PROTECT(names = allocVector(STRSXP, 3));
    SET_STRING_ELT(names, 0, mkChar("first"));
    SET_STRING_ELT(names, 1, mkChar("after"));
    SET_STRING_ELT(names, 2, mkChar("value"));
    namesgets(z, names);

    PROTECT(first = allocVector(INTSXP, z_size));
    PROTECT(after = allocVector(INTSXP, z_size));
    PROTECT(value = allocVector(REALSXP, z_size));
    for(int i=0; i<z_size; i++){
	//Rprintf("%d %d %f\n", z_first[i], z_after[i], z_value[i]);
	INTEGER(first)[i]  = z_first[i];
	INTEGER(after)[i]  = z_after[i];
	REAL(value)[i] = z_value[i];
    }
    SET_VECTOR_ELT(z, 0, first);
    SET_VECTOR_ELT(z, 1, after);
    SET_VECTOR_ELT(z, 2, value);

    UNPROTECT(5);
    return z;
}


R/SparseSignal.R
SparseSignal <- structure(function
### Make a SparseSignal, which is a piecewise constant function
### defined on 1, ..., L. 
(first,
### Integer vector: first position of each segment.
 after,
### Integer vector: first position after each segment.
 value
### Numeric vector: segment values.
 ){
  stopifnot(is.numeric(first))
  stopifnot(is.numeric(after))
  stopifnot(is.numeric(value))
  stopifnot(first < after)
  stopifnot(c(first, after) > 0)
  ss <- data.frame(first=as.integer(first),
                   after=as.integer(after),
                   value)[value != 0,]
  class(ss) <- c("SparseSignal", "data.frame")
  ss
### data.frame with columns first, after, value. 1 row per segment,
### and there are no rows for segments with value 0.
}, ex=function(){
  ## SparseSignals do not have 0-valued segments.
  ss <- SparseSignal(c(10, 12, 15, 1000),
                     c(11, 13, 17, 1001),
                     c( 1, -2,  4,    0))
  print(ss)
})



src/SparseSignals.c
#include <stdio.h>
#include "SparseSignals.h"

int add_SparseSignals(
    int *x_first, int *x_after, double *x_value, int x_size,
    int *y_first, int *y_after, double *y_value, int y_size,
    int *z_first, int *z_after, double *z_value){
    int x_start=1, x_pos, y_start=1, y_pos, z_size=0, e_pos,
	last_pos=0, new_first, new_after, last_x_pos=0, last_y_pos=0;
    double e_value, value=0;
    while(x_size || y_size){
	if(x_start){
	    x_pos = *x_first;
	}else{
	    x_pos = *x_after;
	}
	if(y_start){
	    y_pos = *y_first;
	}else{
	    y_pos = *y_after;
	}
	// make sure they are never the smallest if there are no more 
	// data left.
	if(x_size == 0){
	    x_pos = y_pos + 1;
	}
	if(y_size == 0){
	    y_pos = x_pos + 1;
	}
	if(x_pos < last_x_pos || y_pos < last_y_pos){
	    /* printf("should be x %d >= %d, y %d >= %d\n",  */
	    /* 	   x_pos, last_x_pos, y_pos, last_y_pos); */
	    return ERROR_POSITIONS_DECREASING;
	}
	last_y_pos = y_pos;
	last_x_pos = x_pos;
	/* printf("x_size=%d y_size=%d x_pos=%d y_pos=%d\n",  */
	/*        x_size, y_size, x_pos, y_pos); */
	if(x_pos < y_pos){
	    e_pos = x_pos;
	    if(x_start){
		e_value = *x_value;
		x_first++;
		x_start = 0;
	    }else{
		e_value = - *x_value;
		x_after++;
		x_start = 1;
		x_size--;
		x_value++;
	    }
	}else{
	    e_pos = y_pos;
	    if(y_start){
		e_value = *y_value;
		y_first++;
		y_start = 0;
	    }else{
		e_value = - *y_value;
		y_after++;
		y_start = 1;
		y_size--;
		y_value++;
	    }
	}
	//printf("%10d=last_pos %10d=e_pos %10.2f=value %10.2f=e_value\n", 
	//     last_pos, e_pos, value, e_value);
	if(last_pos != e_pos){
	    if(value != 0){
		//printf("%10.2f=value pos change %5d to %5d\n", 
		//     value, last_pos, e_pos);
		z_value[z_size] = value;
		z_first[z_size] = last_pos;
		z_after[z_size] = e_pos;
		z_size++;
	    }
	}
	value += e_value;
	last_pos = e_pos;
    }
    return z_size;
}


R/methods.R
### Test if x is a SparseSignal.
is.SparseSignal <- function(x){
  "SparseSignal" %in% class(x)
### TRUE or FALSE.
}

### Add two SparseSignals.
"+.SparseSignal" <- structure(function(x, y){
  if(is.SparseSignal(y)){
    L <- addSparseSignals(x, y)
    with(L, SparseSignal(first, after, value))
  }else if(is.numeric(y)){
    if(length(y) %in% c(1, nrow(x))){
      x$value <- x$value + y
      x[x$value != 0,]
    }else{
      stop("only can add numeric vectors of size 1 or nrow(x)")
    }
  }else{
    stop("addition only defined when y is numeric or SparseSignal")
  }
}, ex=function(){
  ## adding two non-overlapping 1-segment signals gives a signal with
  ## 2 segments.
  x <- SparseSignal(1, 2, 5.5)
  x + SparseSignal(4, 10, 3.8) # 2 rows.
  ## adding two overlapping 1-segment signals with opposite values
  ## gives a completely sparse signal with 0 segments.
  none <- x + SparseSignal(1, 2, -5.5) # 0 rows.
  str(none)
  ## a more complicated example.
  x <- SparseSignal(c(10L, 12L, 14L),
                    c(11L, 13L, 16L),
                    c(  1,   2,   3))
  y <- SparseSignal(c(10L, 12L, 15L),
                    c(11L, 13L, 17L),
                    c( 1, -2,    4))
  x + y
  ## you can also just add vectors, but they need to be length 1 or
  ## the number of segments.
  x + 1000
  x + c(1, 200, 3000)
})

### Subtract two SparseSignals.
"-.SparseSignal" <- structure(function(x, y){
  if(is.SparseSignal(y)){
    y$value <- -y$value
    x + y
  }else if(is.numeric(y)){
    x + (-y)
  }else{
    stop("subtraction only defined when y is numeric or SparseSignal")
  }
}, ex=function(){
  ## subtracting two non-overlapping 1-segment signals gives a signal
  ## with 2 segments.
  x <- SparseSignal(1, 2, 5.5)
  x - SparseSignal(4, 10, 3.8) # 2 rows.
  ## subtracting a SparseSignal from itself gives an empty signal.
  none <- x - x
  str(none)
})

### Print a summary of the SparseSignal.
print.SparseSignal <- function(x, ...){
  nrows <- getOption("nrowSparseSignal")
  if(is.null(nrows))nrows <- 10
  cat(sprintf("SparseSignal with %d nonzero segments from %d to %d\n",
              nrow(x), x$first[1], x$after[nrow(x)]-1))
  print.data.frame(head(x, nrows))
  if(nrow(x) > nrows){
    cat("...\n")
  }
  invisible(x)
}

### Get a piecewise constant vector from first up to but not including
### after. This is inefficient and should be used only to compare
### speed with efficient operations.
getVector <- structure(function(ss, first, after){
  first <- as.integer(first)
  after <- as.integer(after)
  stopifnot(length(first)==1)
  stopifnot(length(after)==1)
  size <- after-first
  vec <- rep(NA, size)
  not.useful <- ss$after <= first & ss$first >= after
  useful <- ss[!not.useful,]
  vec.i <- 1 #next position to write to.
  ss.i <- first #next position in ss to read from.
  for(seg.i in 1:nrow(useful)){
    seg.first <- useful$first[seg.i]
    if(seg.first > ss.i){
      zero.size <- seg.first-ss.i
      vec.i.end <- vec.i+zero.size-1
      vec[vec.i:vec.i.end] <- 0
      vec.i <- vec.i.end+1
      ss.i <- seg.first
    }
    seg.after <- useful$after[seg.i]
    seg.size <- seg.after-seg.first
    vec.i.end <- vec.i+seg.size-1
    vec[vec.i:vec.i.end] <- useful$value[seg.i]
    vec.i <- vec.i.end+1
    ss.i <- seg.after
  }
  if(vec.i <= length(vec)){
    vec[vec.i:length(vec)] <- 0
  }
  vec
}, ex=function(){
  x <- SparseSignal(c(10, 15), c(13, 20), c(3.3, 5.5))
  show <- function(first, after){
    v <- getVector(x, first, after)
    data.frame(v, i=first:(after-1))
  }
  show(5, 25)
  show(10, 20)
  show(10, 21)
  show(9, 21)
})
